### PR TITLE
[Merged by Bors] - feat: Port/data.list.nat antidiagonal

### PR DESCRIPTION
**Pull request by casavaca**
---

WIP

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/List/NatAntidiagonal.lean [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+# Antidiagonals in ℕ × ℕ as lists
+
+This file defines the antidiagonals of ℕ × ℕ as lists: the `n`-th antidiagonal is the list of
+pairs `(i, j)` such that `i + j = n`. This is useful for polynomial multiplication and more
+generally for sums going from `0` to `n`.
+
+## Notes
+
+Files `data.multiset.nat_antidiagonal` and `data.finset.nat_antidiagonal` successively turn the
+`list` definition we have here into `multiset` and `finset`.
```
Comment (jcommelin) [inline Mathlib/Data/List/NatAntidiagonal.lean:24]:
```suggestion
Files `Data.Multiset.NatAntidiagonal` and `Data.Finset.NatAntidiagonal` successively turn the
`List` definition we have here into `Multiset` and `Finset`.
```

File: Mathlib/Data/List/NatAntidiagonal.lean [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+/-- The antidiagonal of a natural number `n` is the list of pairs `(i, j)` such that `i + j = n`. -/
+def antidiagonal (n : ℕ) : List (ℕ × ℕ) :=
+  (range (n + 1)).map fun i ↦ (i, n - i)
+#align list.nat.antidiagonal List.Nat.antidiagonal
+
+/-- A pair (i, j) is contained in the antidiagonal of `n` if and only if `i + j = n`. -/
+@[simp]
+theorem mem_antidiagonal {n : ℕ} {x : ℕ × ℕ} : x ∈ antidiagonal n ↔ x.1 + x.2 = n :=
+  by
```
Comment (jcommelin) [inline Mathlib/Data/List/NatAntidiagonal.lean:42]:
```suggestion
theorem mem_antidiagonal {n : ℕ} {x : ℕ × ℕ} : x ∈ antidiagonal n ↔ x.1 + x.2 = n := by
```

File: Mathlib/Data/List/NatAntidiagonal.lean [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+/-- The antidiagonal of `n` does not contain duplicate entries. -/
+theorem nodup_antidiagonal (n : ℕ) : Nodup (antidiagonal n) :=
+  (nodup_range _).map ((@LeftInverse.injective ℕ (ℕ × ℕ) Prod.fst fun i ↦ (i, n - i)) fun _ ↦ rfl)
+#align list.nat.nodup_antidiagonal List.Nat.nodup_antidiagonal
+
+@[simp]
+theorem antidiagonal_succ {n : ℕ} :
+    antidiagonal (n + 1) = (0, n + 1) :: (antidiagonal n).map (Prod.map Nat.succ id) :=
+  by
```
Comment (jcommelin) [inline Mathlib/Data/List/NatAntidiagonal.lean:74]:
```suggestion
    antidiagonal (n + 1) = (0, n + 1) :: (antidiagonal n).map (Prod.map Nat.succ id) := by
```

File: Mathlib/Data/List/NatAntidiagonal.lean [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  simp only [antidiagonal, range_succ_eq_map, map_cons, true_and_iff, Nat.add_succ_sub_one,
+    add_zero, id.def, eq_self_iff_true, tsub_zero, map_map, Prod.map_mk]
+  apply congr rfl (congr rfl _)
+  ext; simp
+#align list.nat.antidiagonal_succ List.Nat.antidiagonal_succ
+
+theorem antidiagonal_succ' {n : ℕ} :
+    antidiagonal (n + 1) = (antidiagonal n).map (Prod.map id Nat.succ) ++ [(n + 1, 0)] :=
+  by
```
Comment (jcommelin) [inline Mathlib/Data/List/NatAntidiagonal.lean:83]:
```suggestion
    antidiagonal (n + 1) = (antidiagonal n).map (Prod.map id Nat.succ) ++ [(n + 1, 0)] := by
```

File: Mathlib/Data/List/NatAntidiagonal.lean [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    singleton_append, map_map, map]
+  congr 1
+  apply map_congr
+  simp (config := { contextual := true }) [le_of_lt, Nat.succ_eq_add_one, Nat.sub_add_comm]
+#align list.nat.antidiagonal_succ' List.Nat.antidiagonal_succ'
+
+theorem antidiagonal_succ_succ' {n : ℕ} :
+    antidiagonal (n + 2) =
+      (0, n + 2) :: (antidiagonal n).map (Prod.map Nat.succ Nat.succ) ++ [(n + 2, 0)] :=
+  by
```
Comment (jcommelin) [inline Mathlib/Data/List/NatAntidiagonal.lean:94]:
```suggestion
      (0, n + 2) :: (antidiagonal n).map (Prod.map Nat.succ Nat.succ) ++ [(n + 2, 0)] := by
```

File: Mathlib/Data/List/NatAntidiagonal.lean [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+  rw [antidiagonal_succ']
+  simp
+  ext
+  simp
+#align list.nat.antidiagonal_succ_succ' List.Nat.antidiagonal_succ_succ'
+
+theorem map_swap_antidiagonal {n : ℕ} : (antidiagonal n).map Prod.swap = (antidiagonal n).reverse :=
+  by
```
Comment (jcommelin) [inline Mathlib/Data/List/NatAntidiagonal.lean:102]:
```suggestion
theorem map_swap_antidiagonal {n : ℕ} : (antidiagonal n).map Prod.swap = (antidiagonal n).reverse := by
```

Comment (casavaca) [reply Mathlib/Data/List/NatAntidiagonal.lean:98]: I left this unchanged because otherwise this line will exceeds 100 char (thus resulting in lint error).

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**